pinot-core/src/main/java/org/apache/pinot/core/operator/query/SelectionOperator.java

```diff
@@ -42,212 +39,168 @@
+  /**
+   * NOTE: The expressions in the transform operator has the sort expressions at the front. The sort sequence is the
+   * deduplicated sort expressions.
+   */
+  public SelectionOperator(IndexSegment indexSegment, Selection selection, TransformOperator transformOperator,
+      List<SelectionSort> sortSequence) {
     _indexSegment = indexSegment;
-    _offset = selection.getOffset();
-    _limit = selection.getSize();
-    _selection = selection;
-    _maxRows = _offset + _limit;
     _transformOperator = transformOperator;
+    _sortSequence = sortSequence;
     _expressions = _transformOperator.getExpressions();
+    _limit = selection.getSize();
 
-    List<String> selectColumns = selection.getSelectionColumns();
-    if (selectColumns.size() == 1 && selectColumns.get(0).equals("*")) {
-      selectColumns = new LinkedList<>(indexSegment.getPhysicalColumnNames());
-    }
-
-    List<String> orderByColumns = new ArrayList<>();
-    if (selection.getSelectionSortSequence() != null) {
-      for (SelectionSort selectionSort : selection.getSelectionSortSequence()) {
-        String expression = selectionSort.getColumn();
-        orderByColumns.add(expression);
-      }
-    }
-    _selectExpressions = new ArrayList<>();
-    _orderByExpressions = new ArrayList<>();
-    _orderByIndices = new ArrayList<>();
-    for (int i = 0; i < _expressions.size(); i++) {
-      TransformExpressionTree expression = _expressions.get(i);
-      if (selectColumns.contains(expression.toString())) {
-        _selectExpressions.add(expression);
-      }
-    }
-
-    for (String orderByColumn : orderByColumns) {
-      for (int i = 0; i < _expressions.size(); i++) {
-        TransformExpressionTree expression = _expressions.get(i);
-        if (orderByColumn.equalsIgnoreCase(expression.toString())) {
-          _orderByExpressions.add(expression);
-          _orderByIndices.add(i);
-        }
-      }
+    boolean selectionOnly = sortSequence.isEmpty();
+    if (selectionOnly) {
+      // Selection only
+      _maxRows = _limit;
+      _rows = new ArrayList<>(_maxRows);
+      _priorityQueue = null;
+    } else {
+      // Selection order-by
+      _maxRows = selection.getOffset() + _limit;
+      _rows = null;
+      _priorityQueue = new PriorityQueue<>(_maxRows, getComparator());
     }
 
-    _blockValSets = new BlockValSet[_expressions.size()];
-    _expressionResultMetadata = new TransformResultMetadata[_expressions.size()];
-    _dictionaries = new Dictionary[_expressions.size()];
-    DataSchema.ColumnDataType[] columnDataTypes = new DataSchema.ColumnDataType[_expressions.size()];
-    String[] columnNames = new String[_expressions.size()];
-    for (int i = 0; i < _expressions.size(); i++) {
+    int numExpressions = _expressions.size();
+    _blockValSets = new BlockValSet[numExpressions];
+    String[] columnNames = new String[numExpressions];
+    ColumnDataType[] columnDataTypes = new ColumnDataType[numExpressions];
```

Comment:
I feel the changes here will break the ordering of columns -- TransformOperator has compiled expression trees as as list but that is given to it by TransformPlanNode which constructs them in a set (thus we lose the order as specified in the query). This is why here in the operator we build the list of expressions in the order as specified in the query and fetch block for each in the same order.

Looks like the new changes (although better that they look at the selection columns only once and directly work on the expressions from TransformOperator), we will probably project the columns in an order different than what was specified by the user in query.

This is the reason why existing implementation was getting the list of expressions from TransformOperator but still constructing its local list based on the ordering of columns in the query